### PR TITLE
feat(provider/amazon): Support add/remove instance from target group

### PR DESCRIPTION
**Pull request by jrsquared**

### CODE REVIEW

File: app/scripts/modules/amazon/src/instance/amazon.instance.write.service.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                     protected providerServiceDelegate: ProviderServiceDelegate) {
+    'ngInject';
+    super(taskExecutor, serverGroupReader, providerServiceDelegate);
+  }
+
+  public deregisterInstancesFromTargetGroup(instanceGroups: IMultiInstanceGroup[], application: Application,
+                                             targetGroupNames: string[]): IPromise<ITask> {
+    const jobs = this.buildMultiInstanceJob(instanceGroups, 'deregisterInstancesFromLoadBalancer') as IAmazonMultiInstanceJob[];
+    jobs.forEach((job) => job.targetGroupNames = targetGroupNames);
+    const descriptor = this.buildMultiInstanceDescriptor(jobs, 'Deregister', `from ${targetGroupNames.join(' and ')}`);
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/instance/amazon.instance.write.service.ts]: I bet this looks great if there are 27 target groups

File: app/scripts/modules/amazon/src/instance/amazon.instance.write.service.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    return this.taskExecutor.executeTask({
+      job: jobs,
+      application: application,
+      description: descriptor,
+    });
+  }
+
+  public deregisterInstanceFromTargetGroup(instance: IAmazonInstance, application: Application, params: any = {}): IPromise<ITask> {
+    params.type = 'deregisterInstancesFromLoadBalancer';
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/instance/amazon.instance.write.service.ts]: Is this right? Does Clouddriver know to look at the `targetGroupNames` to figure out what to do or something?

File: app/scripts/modules/amazon/src/instance/amazon.instance.write.service.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    params.cloudProvider = instance.cloudProvider;
+    return this.taskExecutor.executeTask({
+      job: [params],
+      application: application,
+      description: `Register instance: ${instance.id}`
+    });
+  }
+}
+
+export const AMAZON_INSTANCE_WRITE_SERVICE = 'spinnaker.aws.instance.write.service';
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/instance/amazon.instance.write.service.ts]: `spinnaker.amazon.instance.write.service`

File: app/scripts/modules/amazon/src/instance/details/instance.details.controller.js [hunk 4]
```diff
@@ -99,6 +101,18 @@ module.exports = angular.module('spinnaker.amazon.instance.details.controller',
                 vpcId = loadBalancer.vpcId;
                 return true;
               }
+            }) ||
+            loadBalancer.targetGroups.some(function (targetGroup) {
+              return targetGroup.instances.some(function (possibleInstance) {
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/instance/details/instance.details.controller.js]: are you familiar with arrow functions?

File: app/scripts/modules/amazon/src/instance/details/instance.details.controller.js [hunk 5]
```diff
@@ -118,6 +132,23 @@ module.exports = angular.module('spinnaker.amazon.instance.details.controller',
                     return true;
                   }
                 });
+              }) ||
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/instance/details/instance.details.controller.js]: seeing the `||` makes me think maybe these matching functions should be split up to make this a little more readable.

(I know it's already messy code, but this is...um...)

File: app/scripts/modules/amazon/src/instance/details/instance.details.controller.js [hunk 6]
```diff
@@ -198,6 +230,28 @@ module.exports = angular.module('spinnaker.amazon.instance.details.controller',
...
+      }
+      var outOfService = healthMetrics.some(function(health) {
+        return health.type === 'TargetGroup' && health.state === 'OutOfService';
+      });
+      var hasTargetGroupHealth = healthMetrics.some(function(health) {
+        return health.type === 'TargetGroup';
+      });
+      return outOfService || !hasTargetGroupHealth;
+    };
+
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/instance/details/instance.details.controller.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions

File: app/scripts/modules/amazon/src/instance/details/instance.details.controller.js [hunk 7]
```diff
@@ -332,6 +386,51 @@ module.exports = angular.module('spinnaker.amazon.instance.details.controller',
       });
     };
 
+    this.registerInstanceWithTargetGroup = function registerInstanceWithTargetGroup() {
+      var instance = $scope.instance;
+      var targetGroupNames = instance.targetGroups.join(' and ');
+
+      var taskMonitor = {
+        application: app,
+        title: 'Registering ' + instance.instanceId + ' with ' + targetGroupNames
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/instance/details/instance.details.controller.js]: template string

File: app/scripts/modules/amazon/src/instance/details/instance.details.controller.js [hunk 8]
```diff
@@ -332,6 +386,51 @@ module.exports = angular.module('spinnaker.amazon.instance.details.controller',
...
+      };
+
+      var submitMethod = function () {
+        return amazonInstanceWriter.registerInstanceWithTargetGroup(instance, app);
+      };
+
+      confirmationModalService.confirm({
+        header: 'Really register ' + instance.instanceId + ' with ' + targetGroupNames + '?',
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/instance/details/instance.details.controller.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

File: app/scripts/modules/amazon/src/instance/details/instance.details.controller.js [hunk 9]
```diff
@@ -332,6 +386,51 @@ module.exports = angular.module('spinnaker.amazon.instance.details.controller',
...
+        application: app,
+        title: 'Deregistering ' + instance.instanceId + ' from ' + targetGroupNames
+      };
+
+      var submitMethod = function () {
+        return amazonInstanceWriter.deregisterInstanceFromTargetGroup(instance, app);
+      };
+
+      confirmationModalService.confirm({
+        header: 'Really deregister ' + instance.instanceId + ' from ' + targetGroupNames + '?',
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/instance/details/instance.details.controller.js]: let me know if you need help converting some of this to modern JS

File: yarn.lock [hunk 10]
```diff
@@ -5975,9 +5975,9 @@ typedarray@^0.0.6:
   version "0.0.6"
   resolved "https://registry.yarnpkg.com/typedarray/-/typedarray-0.0.6.tgz#867ac74e3864187b1d3d47d996a78ec5c8830777"
 
-typescript@^2.4.1:
-  version "2.4.1"
-  resolved "https://registry.yarnpkg.com/typescript/-/typescript-2.4.1.tgz#c3ccb16ddaa0b2314de031e7e6fee89e5ba346bc"
+typescript@~2.4.1:
+  version "2.4.2"
+  resolved "https://registry.yarnpkg.com/typescript/-/typescript-2.4.2.tgz#f8395f85d459276067c988aa41837a8f82870844"
```
Comment (anotherchrisberry) [inline yarn.lock]: hmmm

File: app/scripts/modules/amazon/src/instance/amazon.instance.write.service.ts (hunk 1 above)
Comment (jrsquared) [inline app/scripts/modules/amazon/src/instance/amazon.instance.write.service.ts]: Bahaha, just as good as when there's 27 load balancers.

File: app/scripts/modules/amazon/src/instance/amazon.instance.write.service.ts (hunk 2 above)
Comment (jrsquared) [inline app/scripts/modules/amazon/src/instance/amazon.instance.write.service.ts]: Yes.

File: app/scripts/modules/amazon/src/instance/details/instance.details.controller.js (hunk 4 above)
Comment (jrsquared) [inline app/scripts/modules/amazon/src/instance/details/instance.details.controller.js]: Yeah, but I didn't want it to feel left out with all the rest of `function()`s. Refactor of this class will come later.

File: yarn.lock (hunk 10 above)
Comment (jrsquared) [inline yarn.lock]: Gotta fix the yarn.lock when it's not committed properly before.

File: app/scripts/modules/amazon/src/instance/amazon.instance.write.service.ts (hunk 2 above)
Comment (jrsquared) [inline app/scripts/modules/amazon/src/instance/amazon.instance.write.service.ts]: For context: We didn't want to add a different type of operation that woudl be amazon specific. so reusing the old type, and it checks if there are `loadBalancerNames` or `targetGroupNames` in the parameters.

**Pull request merged**